ovn_k8s/modes/overlay.py

```diff
@@ -304,3 +312,124 @@ def add_endpoint(self, event):
+                    ovn_nbctl("--if-exists", "lsp-del", logical_port)
+                except Exception as e:
+                    vlog.err("sync_pods: failed to delete logical_port %s"
+                             % (logical_port))
+                vlog.info("sync_pods: Deleted logical port %s"
+                          % (logical_port))
+
+    def _get_load_balancer_vips(self, load_balancer):
+        try:
+            vips = ovn_nbctl("--data=bare", "--no-heading", "get",
+                             "load_balancer", load_balancer,
+                             "vips").replace('=', ":")
+            return ast.literal_eval(vips)
+        except Exception as e:
+            vlog.err("_get_load_balancer_vips: failed to get vips for %s (%s)"
+                     % (load_balancer, str(e)))
+            return None
+
+    def sync_services(self, services):
+        # For all the services, we will populate the below lists with
+        # IP:port that act as VIP in the OVN load-balancers.
+        tcp_nodeport_services = []
+        udp_nodeport_services = []
+        tcp_services = []
+        udp_services = []
+        if services and 'items' in services:
+            for service in services['items']:
+                service_type = service['spec'].get('type')
+                if service_type != "ClusterIP" and service_type != "NodePort":
+                    continue
+
+                service_ip = service['spec'].get('clusterIP')
+                if not service_ip:
+                    continue
+
+                service_ports = service['spec'].get('ports')
+                if not service_ports:
+                    continue
+
+                external_ips = service['spec'].get('externalIPs')
+
+                for service_port in service_ports:
+                    if service_type == "NodePort":
+                        port = service_port.get('nodePort')
+                    else:
+                        port = service_port.get('port')
+
+                    if not port:
+                        continue
+
+                    protocol = service_port.get('protocol', 'TCP')
+
+                    if service_type == "NodePort":
+                        physical_gateway_ips = self._get_physical_gateway_ips()
+                        for gateway_ip in physical_gateway_ips:
+                            key = "%s:%s" % (gateway_ip, port)
+                            if protocol == "TCP":
+                                tcp_nodeport_services.append(key)
+                            else:
+                                udp_nodeport_services.append(key)
+                    elif external_ips:
+                        for external_ip in external_ips:
+                            key = "%s:%s" % (external_ip, port)
+                            if protocol == "TCP":
+                                tcp_nodeport_services.append(key)
+                            else:
+                                udp_nodeport_services.append(key)
```

Comment:
I am not very familiar of doing it the right way using closures. Please feel free to send an incremental later.